// ngx-clipboard/bundles/ngx-clipboard.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/platform-browser'), require('ngx-window-token'), require('@angular/common'), require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('ngx-clipboard', ['exports', '@angular/platform-browser', 'ngx-window-token', '@angular/common', '@angular/core'], factory) :
    (factory((global['ngx-clipboard'] = {}),global.ng.platformBrowser,global.i2,global.ng.common,global.ng.core));
}(this, (function (exports,i1,i2,common,i0) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    // The following code is heavily copy from https://github.com/zenorocha/clipboard.js
    var ClipboardService = /** @class */ (function () {
        function ClipboardService(document, window) {
            this.document = document;
            this.window = window;
        }
        Object.defineProperty(ClipboardService.prototype, "isSupported", {
            get: /**
             * @return {?}
             */ function () {
                return !!this.document.queryCommandSupported && !!this.document.queryCommandSupported('copy') && !!this.window;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} element
         * @return {?}
         */
        ClipboardService.prototype.isTargetValid = /**
         * @param {?} element
         * @return {?}
         */
            function (element) {
                if (element instanceof HTMLInputElement || element instanceof HTMLTextAreaElement) {
                    if (element.hasAttribute('disabled')) {
                        throw new Error('Invalid "target" attribute. Please use "readonly" instead of "disabled" attribute');
                    }
                    return true;
                }
                throw new Error('Target should be input or textarea');
            };
        /**
         * copyFromInputElement
         */
        /**
         * copyFromInputElement
         * @param {?} targetElm
         * @return {?}
         */
        ClipboardService.prototype.copyFromInputElement = /**
         * copyFromInputElement
         * @param {?} targetElm
         * @return {?}
         */
            function (targetElm) {
                try {
                    this.selectTarget(targetElm);
                    /** @type {?} */
                    var re = this.copyText();
                    this.clearSelection(targetElm, this.window);
                    return re && this.isCopySuccessInIE11();
                }
                catch (error) {
                    return false;
                }
            };
        // this is for IE11 return true even if copy fail
        // this is for IE11 return true even if copy fail
        /**
         * @return {?}
         */
        ClipboardService.prototype.isCopySuccessInIE11 =
            // this is for IE11 return true even if copy fail
            /**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var clipboardData = this.window['clipboardData'];
                if (clipboardData && clipboardData.getData) {
                    if (!clipboardData.getData('Text')) {
                        return false;
                    }
                }
                return true;
            };
        /**
         * Creates a fake textarea element, sets its value from `text` property,
         * and makes a selection on it.
         */
        /**
         * Creates a fake textarea element, sets its value from `text` property,
         * and makes a selection on it.
         * @param {?} content
         * @param {?=} container
         * @return {?}
         */
        ClipboardService.prototype.copyFromContent = /**
         * Creates a fake textarea element, sets its value from `text` property,
         * and makes a selection on it.
         * @param {?} content
         * @param {?=} container
         * @return {?}
         */
            function (content, container) {
                if (container === void 0) {
                    container = this.window.document.body;
                }
                // check if the temp textarea is still belong the current container.
                // In case we have multiple places using ngx-clipboard, one is in a modal using container but the other one is not.
                if (this.tempTextArea && !container.contains(this.tempTextArea)) {
                    this.destroy(this.tempTextArea.parentElement);
                }
                if (!this.tempTextArea) {
                    this.tempTextArea = this.createTempTextArea(this.document, this.window);
                    try {
                        container.appendChild(this.tempTextArea);
                    }
                    catch (error) {
                        throw new Error('Container should be a Dom element');
                    }
                }
                this.tempTextArea.value = content;
                return this.copyFromInputElement(this.tempTextArea);
            };
        // remove temporary textarea if any
        // remove temporary textarea if any
        /**
         * @param {?=} container
         * @return {?}
         */
        ClipboardService.prototype.destroy =
            // remove temporary textarea if any
            /**
             * @param {?=} container
             * @return {?}
             */
            function (container) {
                if (container === void 0) {
                    container = this.window.document.body;
                }
                if (this.tempTextArea) {
                    container.removeChild(this.tempTextArea);
                    // removeChild doesn't remove the reference from memory
                    this.tempTextArea = undefined;
                }
            };
        // select the target html input element
        // select the target html input element
        /**
         * @private
         * @param {?} inputElement
         * @return {?}
         */
        ClipboardService.prototype.selectTarget =
            // select the target html input element
            /**
             * @private
             * @param {?} inputElement
             * @return {?}
             */
            function (inputElement) {
                inputElement.select();
                inputElement.setSelectionRange(0, inputElement.value.length);
                return inputElement.value.length;
            };
        /**
         * @private
         * @return {?}
         */
        ClipboardService.prototype.copyText = /**
         * @private
         * @return {?}
         */
            function () {
                return this.document.execCommand('copy');
            };
        // Moves focus away from `target` and back to the trigger, removes current selection.
        // Moves focus away from `target` and back to the trigger, removes current selection.
        /**
         * @private
         * @param {?} inputElement
         * @param {?} window
         * @return {?}
         */
        ClipboardService.prototype.clearSelection =
            // Moves focus away from `target` and back to the trigger, removes current selection.
            /**
             * @private
             * @param {?} inputElement
             * @param {?} window
             * @return {?}
             */
            function (inputElement, window) {
                // tslint:disable-next-line:no-unused-expression
                inputElement && inputElement.focus();
                window.getSelection().removeAllRanges();
            };
        // create a fake textarea for copy command
        // create a fake textarea for copy command
        /**
         * @private
         * @param {?} doc
         * @param {?} window
         * @return {?}
         */
        ClipboardService.prototype.createTempTextArea =
            // create a fake textarea for copy command
            /**
             * @private
             * @param {?} doc
             * @param {?} window
             * @return {?}
             */
            function (doc, window) {
                /** @type {?} */
                var isRTL = doc.documentElement.getAttribute('dir') === 'rtl';
                /** @type {?} */
                var ta;
                ta = doc.createElement('textarea');
                // Prevent zooming on iOS
                ta.style.fontSize = '12pt';
                // Reset box model
                ta.style.border = '0';
                ta.style.padding = '0';
                ta.style.margin = '0';
                // Move element out of screen horizontally
                ta.style.position = 'absolute';
                ta.style[isRTL ? 'right' : 'left'] = '-9999px';
                // Move element to the same position vertically
                /** @type {?} */
                var yPosition = window.pageYOffset || doc.documentElement.scrollTop;
                ta.style.top = yPosition + 'px';
                ta.setAttribute('readonly', '');
                return ta;
            };
        ClipboardService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        ClipboardService.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: i0.Inject, args: [i1.DOCUMENT,] }] },
                { type: undefined, decorators: [{ type: i0.Inject, args: [i2.WINDOW,] }] }
            ];
        };
        /** @nocollapse */ ClipboardService.ngInjectableDef = i0.defineInjectable({ factory: function ClipboardService_Factory() { return new ClipboardService(i0.inject(i1.DOCUMENT), i0.inject(i2.WINDOW)); }, token: ClipboardService, providedIn: "root" });
        return ClipboardService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ClipboardDirective = /** @class */ (function () {
        function ClipboardDirective(clipboardSrv) {
            this.clipboardSrv = clipboardSrv;
            this.cbOnSuccess = new i0.EventEmitter();
            this.cbOnError = new i0.EventEmitter();
        }
        // tslint:disable-next-line:no-empty
        // tslint:disable-next-line:no-empty
        /**
         * @return {?}
         */
        ClipboardDirective.prototype.ngOnInit =
            // tslint:disable-next-line:no-empty
            /**
             * @return {?}
             */
            function () { };
        /**
         * @return {?}
         */
        ClipboardDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.clipboardSrv.destroy(this.container);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ClipboardDirective.prototype.onClick = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (!this.clipboardSrv.isSupported) {
                    this.handleResult(false, undefined, event);
                }
                else if (this.targetElm && this.clipboardSrv.isTargetValid(this.targetElm)) {
                    this.handleResult(this.clipboardSrv.copyFromInputElement(this.targetElm), this.targetElm.value, event);
                }
                else if (this.cbContent) {
                    this.handleResult(this.clipboardSrv.copyFromContent(this.cbContent, this.container), this.cbContent, event);
                }
            };
        /**
         * Fires an event based on the copy operation result.
         * @param succeeded
         */
        /**
         * Fires an event based on the copy operation result.
         * @private
         * @param {?} succeeded
         * @param {?} copiedContent
         * @param {?} event
         * @return {?}
         */
        ClipboardDirective.prototype.handleResult = /**
         * Fires an event based on the copy operation result.
         * @private
         * @param {?} succeeded
         * @param {?} copiedContent
         * @param {?} event
         * @return {?}
         */
            function (succeeded, copiedContent, event) {
                if (succeeded) {
                    this.cbOnSuccess.emit({ isSuccess: true, content: copiedContent, event: event });
                }
                else {
                    this.cbOnError.emit({ isSuccess: false, event: event });
                }
            };
        ClipboardDirective.decorators = [
            { type: i0.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: '[ngxClipboard]'
                    },] }
        ];
        /** @nocollapse */
        ClipboardDirective.ctorParameters = function () {
            return [
                { type: ClipboardService }
            ];
        };
        ClipboardDirective.propDecorators = {
            targetElm: [{ type: i0.Input, args: ['ngxClipboard',] }],
            container: [{ type: i0.Input }],
            cbContent: [{ type: i0.Input }],
            cbOnSuccess: [{ type: i0.Output }],
            cbOnError: [{ type: i0.Output }],
            onClick: [{ type: i0.HostListener, args: ['click', ['$event.target'],] }]
        };
        return ClipboardDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ClipboardModule = /** @class */ (function () {
        function ClipboardModule() {
        }
        ClipboardModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [common.CommonModule],
                        declarations: [ClipboardDirective],
                        exports: [ClipboardDirective]
                    },] }
        ];
        return ClipboardModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.ClipboardService = ClipboardService;
    exports.ClipboardDirective = ClipboardDirective;
    exports.ClipboardModule = ClipboardModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
